### ISSUE TITLE
Reading TemSen in the Android app

### ISSUE THREAD

**Issue by satcom886** [tomikaa87/gree-remote#31]
It looks like some ACs report current room temperature with the `TemSen` property. The reported value has `+40°C` offset. Would this be possible to integrate into your programs?

I should also add that my Sinclair AC with firmware `3.7` didn't report the temperature until I "updated" to `2.5`. Before that it just reported 0.

**Comment by tomikaa87** [comment]
Hi @satcom886,

I have some ongoing projects at the moment, but in the near future I can allocate some time to add it to the Android app.

**Comment by alex-013** [comment]
Hi, @satcom886 how did you "update" your firmaware? When i open the app and click on firmware update i get current version `3.7` server version `3.42` and back button.

**Comment by satcom886** [comment]
I mentioned that [here](https://github.com/inwaar/gree-hvac-client/pull/11#issuecomment-652491430).

**Comment by xuefer** [comment]
there's no +40°C offset reported by V4 from gree+ app (iOS)

```
gree-remote/PythonCLI
# ./get TemSen
gree.py -c 192.168.26.72 -i 77a8ba4a -k ... get TemSen
Getting parameter: TemSen
TemSen = 25

gree.py -c 192.168.26.73 -i 77dd30f3 -k ... get TemSen
Getting parameter: TemSen
TemSen = 26

gree.py -c 192.168.26.74 -i a70f1fa6 -k ... get TemSen
Getting parameter: TemSen
TemSen = 26

gree.py -c 192.168.26.75 -i a7102045 -k ... get TemSen
Getting parameter: TemSen
TemSen = 24

```

**Comment by velmafia** [comment]
@xuefer how do you upgrade firmware to v4?

**Comment by xuefer** [comment]
> @xuefer how do you upgrade firmware to v4?

just by the official app, the China mainlain version from China area of app store which is accessible with China apple account(The apple account can't switch area once registered). I'm not sure if it's triggered the app version or by your device model though

**Comment by velmafia** [comment]
Thanks for answer, in Europe region (and Europe A/C distribution) I don't see any update in android app. In next days I will create new gree account, and chose China region to check what will happen.

**Comment by velmafia** [comment]
I tried today firmware upgrade with China region for Gree+ android app (new gree account and A/C registration
So for my Gree Luxury Plus:
- Europe region: available only `3.42` firmware
- China region: I was able to upgrade to `3.70`
- by `WiFi Smart` app, `2.5` firmware is available.

Version and `TemSen` summary:
 - `2.5`: return current temperature with +40 offset
 - `3.42`: return current temperature with +40 offset
 - `3.70`: return always 0
 - `v4`: return correct temperature - [tested](https://github.com/tomikaa87/gree-remote/issues/31#issuecomment-734716981) by @xuefer

**Comment by dulemis** [comment]
I have gree amber 25 with firmware 3.67. TemSen returns +40 offset.

**Comment by dulemis** [comment]
I just noticed on my unit that the returned temperature is different depending on if I switch between cooling and heating!? It's a 3c difference. Could someone confirm?

**Comment by aivus** [comment]
@dulemis I can confirm this and I see the same temperature change on internal AC display.
I guess it was added by developers to add some temperature difference between different modes.

**Comment by dulemis** [comment]
> @dulemis I can confirm this and I see the same temperature change on internal AC display.
> I guess it was added by developers to add some temperature difference between different modes.

So you don't think this is a bug? Can't really see what would be the benefit. Which temperature is the correct one then...?

**Comment by aivus** [comment]
Yes, I don't think that this is a bug.

Probably this change is compensating heating/cooling of the temp sensor itself during the work. But it's just my guess.

**Comment by satcom886** [comment]
@aivus Yes, that's also what I think. It's not a very good solution (or really not a solution at all), but I guess their devs at least tried to compensate for the heating.

**Comment by dulemis** [comment]
I don't quite understand, why is there need for compensation? I mean, the room temperature is still the same obviously..

**Comment by dulemis** [comment]
Actually, the temperature returned when swtiching to cooling is more "correct" in my house then the other. (compared with a digital temperature unit on the wall). It says right now 22 c, while the AC unit says 20 c in heating-mode and 23c in cooling-mode. So it's really weird, which one it is in fact and why it differs 3 degres as "baseline"...

**Comment by dulemis** [comment]
Just upgraded to v 3.72 and the TemSen returns +40 offret (not 0).

**Comment by xuefer** [comment]
> How did you update the firmware? Because the iOS app does not have firmware update function.

3 dots on the top-right corner, Device Information, Firmware Version. Update button will be available only when new version is available.

There's no version upgrade for my AC model. too bad, i really need it fixed to keep updating TemSen

**Comment by zsiraph** [comment]
Thanks, I found it, and looks like 3.72 fixed my problem (-40C all day everyday :D ). 
The UX of the official app is horrendous.

**Comment by aivus** [comment]
Just an interesting findings.

Several days ago one user reported issues with current temp value. I checked the logs and found this:
```
[finder] Message:  {
  t: 'dev',
  cid: 'xxxxxxxxx',
  bc: 'gree',
  brand: 'gree',
  catalog: 'gree',
  mac: 'xxxxxxxxx',
  mid: '10002',
  model: 'gree',
  name: 'xxxxxxxxx',
  series: 'gree',
  vender: '1',
  ver: 'V1.1.13',
  lock: 0
}
...
IN.MSG.UNPACK {
  t: 'dat',
  mac: 'xxxxxxxxx',
  r: 200,
  cols: [
    'Pow',    'Mod',
    'TemUn',  'SetTem',
    'TemSen', 'WdSpd',
    'Air',    'Blo',
    'Health', 'SwhSlp',
    'Lig',    'SwingLfRig',
    'SwUpDn', 'Quiet',
    'Tur',    'SvSt'
  ],
  dat: [
    1, 4, 0, 20, 218, 1,
    0, 0, 0,  0,   1, 0,
    0, 2, 0,  0
  ]
}
```

Meaning `TemSen` always returns `218`. 
And seems it's constant and not correlates with real temperature in the room
The temp sensor in AC presents, because it shows correct temperature on built-in screen when using Temp button)
